packages/amplify_auth_cognito/android/src/main/kotlin/com/amazonaws/amplify/amplify_auth_cognito/AuthCognito.kt

```diff
@@ -156,6 +162,11 @@ public class AuthCognito : FlutterPlugin, ActivityAware, MethodCallHandler, Plug
       return
     }
 
+    if (DeviceHandler.canHandle(call.method)) {
```

Comment:
Is there a device-api-specific reason to create this additional handler, other than the AtomicResult and CoroutineScope stuff you are doing? Or is this just general separation of concerns?

Reply:
General separation and smaller class size, but also the argument handling right below this doesn't allow for `null` to be passed, which I wanted to do.

I also thought it could be cool as a potential refactor option, since if all the handlers were separated out into their own classes, we could keep this class pretty slim:

```kotlin
private val handlers = listOf(DeviceHandler(), ...)

override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
  for (handler in handlers) {
    if (handler.canHandle(call.method)) {
      handler.handle(call, result)
      return
    }
  }
  
  result.notImplemented()
}
```

Reply:
We can change the pattern - I'm not at all attached. For this PR, I needed to separate it out for my own sanity haha. I agree that the use of classes is not totally necessary, although I do like it for the namespacing aspect. But they could all be top-level extension methods on a big class which could be separated out into files, as needed.

I'll leave the null-check up to you - I didn't want to touch previous code as much as possible.

Reply:
I would lean towards leaving the previous code untouched, as we probably just need to refactor that anyway. I like the idea of separate classes for different handlers that this is establishing, so maybe continue to have it handle null in it's own way.  

Regarding the general idea of distinct class for handlers, how would you envision this being applied to the other method calls? What is the criteria that makes these new methods get handled by a single handler, other than the fact that they are being implemented together? Is it the fact that they are part of a sort of common workflow? If that's the case, would it makes sense to implement this pattern in a refactor and group APIs such as `signUp`, `confirmSignUp` and `resendSignUpCode` together?

Reply:
Yeah, like Jordan was saying, the lines are not so clear. It could be like `SignUpHandler` for the methods you mention, then `SignInHandler`, `SignOutHandler`, etc. Some may only have one method. Or it's just all top-level methods/extension methods on this class. The only problem with that is that we still have a long when/case which is hard to reason about.

Reply:
Ah ok - sorry I missed @Jordan-Nelson 's comment.